### PR TITLE
GH-122616: Simplify LOAD_ATTR_WITH_HINT and STORE_ATTR_WITH_HINT

### PR DESCRIPTION
**Pull request by markshannon**
Removes branching code, using a DEOPT.

* Issue: gh-122616

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2214,23 +2208,13 @@ dummy_func(
...
-                ep->me_value = PyStackRef_AsPyObjectSteal(value);
-            }
-            Py_DECREF(old_value);
+            DEOPT_IF(!DK_IS_UNICODE(dict->ma_keys));
+            PyDictUnicodeEntry *ep = DK_UNICODE_ENTRIES(dict->ma_keys) + hint;
+            DEOPT_IF(ep->me_key != name);
+            old_value = ep->me_value;
+            new_version = _PyDict_NotifyEvent(tstate->interp, PyDict_EVENT_MODIFIED, dict, name, PyStackRef_AsPyObjectBorrow(value));
+            ep->me_value = PyStackRef_AsPyObjectSteal(value);
+            Py_XDECREF(old_value);
```
Comment (iritkatriel) [inline Python/bytecodes.c:2218]: You dropped the ``DEOPT_IF(old_value == NULL);``. Doesn't this change semantics?

Comment (markshannon) [reply Python/bytecodes.c:2218]: No. The semantics are the same, just the code path is different.
If the `old_value` is NULL, the `DEOPT` would have resulted in `STORE_ATTR` doing the store. With this PR it is done here.

Comment (markshannon) [reply Python/bytecodes.c:2218]: Actually, it does change things. If `old_value` is `NULL` we should be passing `PyDict_EVENT_ADDED` to `_PyDict_NotifyEvent`.
I'll fix that.

**Review by iritkatriel — APPROVED** [review]

**Comment by markshannon** [comment]
The three failures are networking issues on the emulated platforms and a hypothesis issues

**Comment by scoder** [comment]
A somewhat mysterious bug report (a failure with a Python exception) that we received in Cython under Py3.14a1 bisects to this change:
https://github.com/cython/cython/issues/6426#issuecomment-2422973468
Apparently, it introduced a behavioural change where some object attributes are set to `None` (or possibly their super class default value, i.e. not set at all) instead of the expected value. Note that this is in plain Python code when running Cython to translate one of its modules, running `python cython.py Cython/Compiler/Parsing.py`

**Comment by markshannon** [comment]
@scoder can you make a new CPython issue for the Cython problem?

I'd be very surprised if this PR was the root cause of the problem, but it clearly is a CPython issue not a Cython one.
Thanks.

**Comment by da-woods** [comment]
Hi both: I've created https://github.com/python/cpython/issues/125868 with a bit of extra debugging detail (I stopped trying to debug at the point that reality and pdb started contradicting each other)

**Pull request merged**